src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.chromium.net.CronetEngine;
+import org.chromium.net.CronetException;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataProviders;
+import org.chromium.net.UrlRequest;
+import org.chromium.net.UrlRequest.Callback;
+import org.chromium.net.UrlResponseInfo;
+
+public class CronetHttpStack extends AsyncHttpStack {
+
+    private final CronetEngine mCronetEngine;
+    private final Executor mCallbackExecutor;
+    private final ByteArrayPool mPool;
+
+    public CronetHttpStack(Context context, Executor callbackExecutor, ByteArrayPool pool) {
+        mCronetEngine = new CronetEngine.Builder(context).build();
+        mCallbackExecutor = callbackExecutor;
+        mPool = pool;
+    }
+
+    public CronetHttpStack(Context context) {
+        this(context, (ByteArrayPool) null);
+    }
+
+    public CronetHttpStack(Context context, Executor callbackExecutor) {
+        this(context, callbackExecutor, null);
+    }
+
+    public CronetHttpStack(Context context, ByteArrayPool pool) {
+        this(
+                context,
+                new ThreadPoolExecutor(
+                        0,
+                        1,
+                        60,
+                        TimeUnit.SECONDS,
+                        new LinkedBlockingQueue<Runnable>(),
+                        new ThreadFactory() {
+                            @Override
+                            public Thread newThread(@NonNull Runnable runnable) {
+                                Thread t = Executors.defaultThreadFactory().newThread(runnable);
+                                t.setName("Volley-CronetCallbackThread");
+                                return t;
+                            }
+                        }),
+                pool);
+    }
+
+    @Override
+    public void executeRequest(
+            final Request<?> request,
+            final Map<String, String> additionalHeaders,
+            final OnRequestComplete callback) {
+        final PoolingByteArrayOutputStream bytesReceived = new PoolingByteArrayOutputStream(mPool);
+        final WritableByteChannel receiveChannel = Channels.newChannel(bytesReceived);
+        final Callback urlCallback =
+                new Callback() {
+                    @Override
+                    public void onRedirectReceived(
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo, String s) {
+                        urlRequest.followRedirect();
+                    }
+
+                    @Override
+                    public void onResponseStarted(
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo) {
+                        int httpStatusCode = urlResponseInfo.getHttpStatusCode();
+                        // what checks should we be making of the status code, if any?
```

Comment:
From my understanding, onFailed is called if there isn't a valid response code. By make sure responses are handled appropriately, do you just mean to skip reading if there is no body?

Reply:
I mostly mean "identically to how the other stacks do it", to the extent that it make sense. That said, I'd tend to err on the side of simplicity. I can't find any code history for why HurlStack explicitly handles things like 204s rather than just passing them along to the caller; if a server wants to return a 204 with content then that's their problem IMO (if Cronet itself even permits it). So maybe start simpler and we can add the explicit response code handling if we need it.

Reply:
sounds good